lib/LaTeXML/Package/authblk.sty.ltxml

```diff
@@ -51,27 +60,41 @@ DefConstructor('\affil[]{}',
     $whatsit->setProperty(mark => ToString($mark));
     return; });
 
-Tag('ltx:note',     afterClose => \&authblkRelocateAffil);
-Tag('ltx:document', afterClose => \&pruneEmptyAffil);
+Tag('ltx:document', afterClose => \&authblkRelocateAffil);
 
 sub authblkRelocateAffil {
-  my ($document, $affilnode) = @_;
-  if (($affilnode->getAttribute('role') || '') eq 'affiliationtext') {
-    if (my $mark = $affilnode->getAttribute('mark')) {
-      # Find authors with same mark.
-      my @authors = $document->findnodes(".//ltx:contact[\@role='affiliationmark' and \@_mark='$mark']");
-      foreach my $author (@authors) {
-        $document->appendClone($author, $affilnode->childNodes);
-        $author->setAttribute(role => 'affiliation'); }
-      $affilnode->parentNode->removeChild($affilnode); } }
+  my ($document)   = @_;
+  my @author_nodes = $document->findnodes(".//ltx:contact[\@role='affiliationmark' and \@_mark]");
+  my @affil_nodes  = $document->findnodes(".//ltx:note[\@role='affiliationtext']");
+  # first, unbind all affiliation text nodes - they get reparented or discarded.
+  my %mark_to_affil = ();
+  for my $affil_node (@affil_nodes) {
+    $affil_node->parentNode->removeChild($affil_node);
+    # if we get conflicting marks -- which we shouldn't
+    # the last node wins
+    if (my $mark = $affil_node->getAttribute('mark')) {
+      $mark_to_affil{$mark} = $affil_node; } }
+  # The most accurate approach here is to go author-by-author,
+  # and attach the affiliations in the *exact order* they were requested
+  # \author[1,7]{A} and \author[7,1]{B} are intended to be presented in that order.
+  for my $author_node (@author_nodes) {
+    my @author_marks = split(/\s*,\s*/, ($author_node->getAttribute('_mark') || ''));
```

Comment:
The contents of this loop could be much simpler (just change the role attribute, then
```
$document->appendClone($author_node,$affil_node->childNodes));
```
(no new node, etc)
Except for this packing of multiple marks on a node; Where does this arise?  It seems counter to my intuition of what a mark is supposed to be.

Reply:
Right, so the `$affil_node->childNodes` contains text nodes and text nodes only.

 What we want are sibling `<contact role="affiliation">` nodes,
 which is what I get by cloning the empty template node in  `$author_node`.

I would have preferred to make the new nodes from scratch, without any cloning, but I repeatedly couldn't find a way to do so without creating tens of lines of boilerplate. But anything that creates this tree would be OK by me.

Reply:
As to multiple marks, `\author[1,7]{A}` would initially create the author node:
```xml
<ltx:contact role='affiliationmark' _mark='1,7'></ltx:contact>
```

at the right position in the document tree. Then the subroutine would parse the `'1,7'` in the attribute and deposit the two respective `<contact role="affiliation">` nodes next to it. And in the end, it will unbind the template `affiliationmark` node.

Reply:
Exactly; that seems to me to be wrong. The `\author` should have created sibling contacts for each mark. Then the rest would be simple.

Reply:
I see, reasonable alternative -- I'll send a commit in a few minutes.

Reply:
Alright, refactored. Seeing the same final XML as in my overview comment. See if you like it :>